lib/unread/reader_scopes.rb

```diff
@@ -0,0 +1,29 @@
+module Unread
+  module Reader
+    module Scopes
+      def join_read_marks(readable)
+        assert_readable(readable)
+
+        joins "LEFT JOIN #{ReadMark.table_name} as read_marks ON read_marks.readable_type = '#{readable.class.base_class.name}'
+               AND (read_marks.readable_id   = #{readable.try(readable.class.primary_key)} OR read_marks.readable_id IS NULL)
+               AND read_marks.user_id        = #{table_name}.#{primary_key}
+               AND read_marks.timestamp     >= '#{readable.try(readable.class.readable_options[:on]).to_s(:db)}'"
+      end
+
+      def have_not_read(readable)
+        join_read_marks(readable).where("read_marks.id IS NULL")
+      end
+
+      def have_read(readable)
+        join_read_marks(readable).where('read_marks.id IS NOT NULL')
+      end
+
+      def with_read_marks_for(readable)
+        join_read_marks(readable).select("#{table_name}.*,
+                                          read_marks.id AS read_mark_id,
+                                          '#{readable.class.base_class.name}' AS read_mark_readable_type,
```

Comment:
Or maybe they are used in the other branch (check mikeki/unread and look
for the branch that says read_marks_are_not_shared or something like that)
(I'm away from my laptop right now)
On May 25, 2015 10:21 AM, "Miguel Angel Cervera" <miguelcerverac@gmail.com>
wrote:

> They are actually used in 'reader.rb'
> On May 25, 2015 10:19 AM, "Georg Ledermann" <notifications@github.com>
> wrote:
>
>> In lib/unread/reader_scopes.rb
>> <https://github.com/ledermann/unread/pull/55#discussion_r30988400>:
>>
>> > +               AND read_marks.user_id        = #{table_name}.#{primary_key}
>> > +               AND read_marks.timestamp     >= '#{readable.try(readable.class.readable_options[:on]).to_s(:db)}'"
>> > +      end
>> > +
>> > +      def have_not_read(readable)
>> > +        join_read_marks(readable).where("read_marks.id IS NULL")
>> > +      end
>> > +
>> > +      def have_read(readable)
>> > +        join_read_marks(readable).where('read_marks.id IS NOT NULL')
>> > +      end
>> > +
>> > +      def with_read_marks_for(readable)
>> > +        join_read_marks(readable).select("#{table_name}.*,
>> > +                                          read_marks.id AS read_mark_id,
>> > +                                          '#{readable.class.base_class.name}' AS read_mark_readable_type,
>>
>> So far as I see, the last two selected attributes (
>> read_mark_readable_type and read_mark_readable_id) are never used. Can
>> we remove them?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/ledermann/unread/pull/55/files#r30988400>.
>>
>